[Config][FrameworkBundle] Generate JSON schema for config

valtzu opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #59603
| License       | MIT

Generate JSON schema in `%kernel.build_dir%/config.schema.json` for YAML config when the container is built to improve DX.

### To discuss:

1. ~~Is there any way to deal with PHP enums (`!php/enum` yaml tag)? At the moment they are just ignored/excluded~~ // just allowed a string `!php/enum ...`
1. ~~Should we include `# $schema: path/to/schema.json` in generated yaml files?~~ probably no / up to Flex etc
1. Any edge-cases I missed?
1. ~~Should we rather bump `symfony/config` version constraint in `symfony/framework-bundle` than check for class existence? This is kinda optional stuff so I thought we don't need to enforce it.~~ already bumped in #59649

valtzu committed:
Generate JSON schema for config

GromNaN commented on src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,158 @@
+            default => null,
+        };
+
+        $removeNulls = fn (array $array) => array_filter($array, fn ($value) => null !== $value);
```
This closure should be a method of this class.

GromNaN commented:
Thank you very much for working on this feature.
I gave it a try and it works really well.

- The per-environment configuration need to be added. This is the same as the current root definition (that need to be a type), with key following this pattern: `when@[a-zA-Z0-9]+`
- The JSON schema file will be different for each environment, because the bundles are not the sames. The JSON schema must allow any additional root keys for `when@` other environments.

Here is the files that are not validated by the generated JSON schema:

### `doctrine.yaml`

Here as some required improvements to pass the Symfony Demo config:
- 
- Some `doctrine` configuration are not recognized.

<img width="793" alt="image" src="https://github.com/user-attachments/assets/83f60b50-1f4e-4c78-b5ea-b4e7fd819c18" />

### `framework.yaml`

- All configuration that [`canBeEnabled`](https://github.com/symfony/symfony/blob/d4566b22d633c74b4567a3484ec40b6177d78b1b/src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php#L244) (like `framework.esi`) must accept a boolean value. All [`equivalentValues`](https://github.com/symfony/symfony/blob/d4566b22d633c74b4567a3484ec40b6177d78b1b/src/Symfony/Component/Config/Definition/BaseNode.php#L347) must be allowed.

### `routing.yaml`

- "Missing required property 'resource'. Since the configurations are merged from multiple files (and defaults) properties cannot be required in a single Yaml file.

### `validator.yaml`

- The [`framework.validation.auto_mapping.<class-name>`](https://github.com/symfony/symfony/blob/d4566b22d633c74b4567a3484ec40b6177d78b1b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php#L1069) value must be an object, but an [empty array](https://github.com/symfony/demo/blob/505825290d9f1a50d8ce043bb8b451359de21d24/config/packages/validator.yaml#L9) is accepted.

valtzu commented:
Thanks for quick review @GromNaN!

Resolved `when@` _(though it does not account for env-specific extensions yet)_, removed `"required": [...]` from the schema, added using `equivalentValues`.

The validator issue still on to-do list, not sure on what basis to allow the empty array – maybe `useAttributeAsKey` :thinking: 

The doctrine issue seems to be caused by [normalization in a closure](https://github.com/doctrine/DoctrineBundle/blob/2.13.x/src/DependencyInjection/Configuration.php#L99-L113), not sure if there's anything to do about it. Should we just always allow extra keys / have `"additionalProperties": true` in the schema (for every subschema) for now, not to cause warnings?

GromNaN commented:
> Resolved `when@` _(though it does not account for env-specific extensions yet)_, removed `"required": [...]` from the schema, added using `equivalentValues`.

Thanks, perfect.

> The validator issue still on to-do list, not sure on what basis to allow the empty array – maybe `useAttributeAsKey` 🤔

I think the JSON schema can be a bit more strict than the config validation rules. But this error may be hiding something else.

> The doctrine issue seems to be caused by [normalization in a closure](https://github.com/doctrine/DoctrineBundle/blob/2.13.x/src/DependencyInjection/Configuration.php#L99-L113), some possible solutions below 👇
> 
> 1. always allow extra keys / have `"additionalProperties": true` in the schema (for every subschema), not to cause false-positives on keys that for whatever reason don't appear in the schema

Arbitrary additional properties are not allowed by the Config component. This would fail the purpose of the JSON schema.

> 2. do some manual override / manually defined schema to cover this special case.

It's the last option if you really can't do any better.

> 3. convert _(dynamically?)_  [the xml schema](https://github.com/doctrine/DoctrineBundle/blob/2.13.x/config/schema/doctrine-1.0.xsd) to json schema _(not sure if it matches yaml one-to-one)_

I'd prefer to be able to generate the xsd rather than use it.

Maybe we don't need to support this for a first version, since there's an alternative syntax (more verbose).

GromNaN commented on src/Symfony/Bundle/FrameworkBundle/CacheWarmer/ConfigSchemaCacheWarmer.php:
```diff
@@ -0,0 +1,106 @@
+
+        if ($this->kernel instanceof Kernel) {
+            /** @var ContainerBuilder $container */
+            $container = \Closure::bind(function (Kernel $kernel) {
+                $containerBuilder = $kernel->getContainerBuilder();
+                $kernel->prepareContainer($containerBuilder);
+
+                return $containerBuilder;
+            }, null, $this->kernel)($this->kernel);
```
The closure seems unnecessarily complicated.
```suggestion
            $container = $kernel->getContainerBuilder();
            $kernel->prepareContainer($container);
```

valtzu replied:
But `getContainerBuilder` is a protected method. This block is copy-paste from [`ConfigBuilderCacheWarmer`](https://github.com/symfony/symfony/blob/7.3/src/Symfony/Bundle/FrameworkBundle/CacheWarmer/ConfigBuilderCacheWarmer.php#L51-L69)

---

I agree it is hackish though

GromNaN commented on src/Symfony/Component/Config/Tests/Definition/JsonSchemaGeneratorTest.php:
```diff
@@ -0,0 +1,36 @@
+    {
+        $configuration = new ExampleConfiguration();
+
+        $tmp = tempnam(sys_get_temp_dir(), 'json-schema-generator');
+        register_shutdown_function(fn () => unlink($tmp));
```
The cleanup of the file should be handled in `TestCase::tearDown`. Example in [`CacheWarmerTest`](https://github.com/symfony/symfony/blob/04551ad54245e0f48206eb05a582996b419c8c1e/src/Symfony/Component/HttpKernel/Tests/CacheWarmer/CacheWarmerTest.php#L21-L29)

GromNaN commented on src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,190 @@
+    {
+        $paramTypes = [];
+
+        $getType = fn ($value) => match (get_debug_type($value)) {
```
This closure is not used.

GromNaN commented on src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,190 @@
+            'int' => $removeNulls(['type' => 'integer', 'minimum' => $node->getMin(), 'maximum' => $node->getMax()]),
+            'float' => $removeNulls(['type' => 'number', 'minimum' => $node->getMin(), 'maximum' => $node->getMax()]),
+            'string' => ['type' => 'string'],
+            'enum' => $removeNulls(['enum' => array_map(fn ($v) => $v instanceof \UnitEnum ? \sprintf('!php/enum %s::%s', $v::class, $v->name) : $v, $node->getValues()) ?: null]),
```
```suggestion
            'enum' => $removeNulls(['enum' => array_map(static fn ($v) => $v instanceof \UnitEnum ? \sprintf('!php/enum %s::%s', $v::class, $v->name) : $v, $node->getValues()) ?: null]),
```

GromNaN commented on src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,190 @@
+            $node instanceof PrototypedArrayNode => 'array_prototype',
+            $node instanceof ArrayNode => 'array',
+            $node instanceof ScalarNode => 'scalar',
+            default => null,
```
I don't think there are any cases that don't fit into the above. I would delete the default to leave an error if the class is not expected.

valtzu replied:
As I understand, it's possible to have a completely custom `NodeInterface` implementation in user land, and if you did that, I think it'd be better to allow any value for that field instead of failing to generate the whole schema

GromNaN commented on src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,190 @@
+        $schema = [
+            'type' => ['object', 'array'],
+            'additionalProperties' => $node->shouldIgnoreExtraKeys(),
+            'maxItems' => 0,
```
This `maxItems` only accepts an empty array? This seems strange. It's for [framework.validation.auto_mapping.<class-name>](https://github.com/symfony/symfony/blob/d4566b22d633c74b4567a3484ec40b6177d78b1b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php#L1069), there is no restriction in the array size.

valtzu replied:
The rationale here was that a non-prototyped array means an object – however we'd still allow empty array and treat it the same as empty object as there is no difference in php.

But you're right, it seems validator has the same issue as Doctrine, there is a bunch of [pre-normalization logic](https://github.com/symfony/symfony/blob/d4566b22d633c74b4567a3484ec40b6177d78b1b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php#L1079-L1102) which makes it impossible to determine the allowed schema.

GromNaN reviewed (changes requested):
I did a quick code review.

valtzu commented:
Did a whole lot of cleanup based on the comments.

I also managed to workaround the Doctrine issue by allowing any type when `$node->getNormalizedTypes()` contains `ExprBuilder::TYPE_ANY`.

This of course means we don't get a schema for those, but I think it's better than false-positives when using example configuration.

Also added a deprecation message to the schema which shows up quite nicely now

![image](https://github.com/user-attachments/assets/44b16321-2c67-4a74-8902-c637c0934d0b)

GromNaN commented:
Awesome, the deprecation message is exactly why I want such schema validation.

GromNaN commented on src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,219 @@
+            $allowNull = $allowNull || \in_array(ExprBuilder::TYPE_NULL, $normalizedTypes);
+            if (\in_array(ExprBuilder::TYPE_ANY, $normalizedTypes)) {
+                // This will make IDEs not complain about configurations containing complex beforeNormalization logic
+                $subSchemas[] = new \ArrayObject();
```
For an empty object, it's more correct to use `stdClass`. You are using the fact that any object that doesn't expose any property is encoded as an empty object.
```suggestion
                $subSchemas[] = new \stdClass();
```

valtzu replied:
~~I'm intentionally using `\ArrayObject` (just like [Symfony Serializer does internally](https://github.com/symfony/serializer/blob/7.2/Normalizer/AbstractObjectNormalizer.php#L231)) so that I can perform array operations on it, f.e. `$schema['default'] = $node->getDefaultValue()` and `\count($subSchema, \COUNT_RECURSIVE)`.~~

~~If `\stdClass` was used instead, I think the logic would be overall more complex.~~
 
Switched to `\stdClass` everywhere

GromNaN commented on src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,219 @@
+            }
+        }
+
+        return $subSchemas ?: [new \ArrayObject()];
```
```suggestion
        return $subSchemas ?: [new \stdClass()];
```

GromNaN commented on src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,219 @@
+        $schema['type'][] = $type;
+    }
+
+    private function getReference(array|\ArrayObject $subSchema, array &$definitions): array|\ArrayObject
```
```suggestion
    private function getReference(array|\stdClass $subSchema, array &$definitions): array|\stdClass
```

valtzu replied:
Using `\stdClass` for schemas everywhere now & thus got rid of `array` by-ref, reasoning in https://github.com/symfony/symfony/pull/59620#issuecomment-2666955942

valtzu commented:
I'm a little stuck with the multi-env support: since the schema is generated inside the kernel, it does not seem possible to generate schemas for other environments. And since the kernel class is not really exposed anywhere, it's not trivial to create multiple instances of the kernel with different envs (not that it would be a great idea anyway).

Best I could think of is
1. always only build schema for the current environment
2. in addition to saving the schema into single file inside env-specific directory, also insert a reference into common file outside env-specific dir

This could look something like
```
var/cache/dev/config.schema.json
var/cache/test/config.schema.json
var/cache/whatever/config.schema.json
var/config.schema.json # references all files above, like {"when@dev": { "$ref": "cache/dev/config.schema.json" } }
```

The downside with this solution is that unless you build the container for those other envs, you also don't get the schema for those. Also, I doubt that many people build the container locally for any env but `dev` & `test`.

---

If someone has a better approach, great!

GromNaN commented on src/Symfony/Bundle/FrameworkBundle/CacheWarmer/ConfigSchemaCacheWarmer.php:
```diff
@@ -0,0 +1,106 @@
+
+    public function isOptional(): bool
+    {
+        return false;
```
I think this is optional. The app doesn't relies on it to run.

valtzu replied:
Fixed

GromNaN reviewed (commented):
We could look at `bundles.php` to find all the environments and bundles used, but that wouldn't cover generic use cases (if bundles are registered directly in Kernel.php, for example).

We could also scan all the directories and find all the classes that implement the `BundleInterface`, but that would require scanning composer autoloader and would take up too many bundles.

I think the idea of generating independent schema files for each env is more realist:
- generate `var/cache/dev/config.schema.json`, `var/cache/test/config.schema.json` and `var/cache/prod/config.schema.json`.
- when a file is generated, create or update the parent `var/cache/config.schema.json` to include the generated fields.

I tried this with PHPStorm, but the file loading doesn't work yet. 
```json
{
  "$schema": "http://json-schema.org/draft-06/schema#",
  "allOf": [
    { "$ref": "./dev/config.schema.json" },
    { "$ref": "./test/config.schema.json" }
  ]
}
```

nicolas-grekas commented:
Instead of one big file, we could generate one file per bundle.
That would solve the problem of `when@` and that would align with the practice of splitting config files per bundle.

GromNaN commented:
> Instead of one big file, we could generate one file per bundle. That would solve the problem of `when@` and that would align with the practice of splitting config files per bundle.

That's a very good idea, and it would solve our issue.

GromNaN commented on src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,237 @@
+    private function createObjectSchema(ArrayNode $node, \stdClass $definitions): \stdClass
+    {
+        $schema = (object) [
+            'type' => 'object',
```
If additional properties are allowed, we loose the validation on property names.
I think they can be disallowed once alternative documentation formats are fully described: https://github.com/symfony/symfony/pull/51273
```suggestion
            'type' => 'object',
            'additionalProperties' => false,
```

GromNaN commented on src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,237 @@
+        if ($node instanceof BaseNode && $normalizedTypes = $node->getNormalizedTypes()) {
+            $allowedExtraValues = array_column($node->getEquivalentValues(), 0);
+            $allowNull = $allowNull || \in_array(ExprBuilder::TYPE_NULL, $normalizedTypes);
+            if (\in_array(ExprBuilder::TYPE_ANY, $normalizedTypes)) {
+                // This will make IDEs not complain about configurations containing complex beforeNormalization logic
+                $subSchemas[] = new \stdClass();
+            }
```
This block also need to be reviewed if we can provide a better description of alternative config trees.

GromNaN commented on src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php:
```diff
@@ -0,0 +1,237 @@
+            if ($node->isDeprecated()) {
+                $schema->deprecated = true;
+                $deprecation = $node->getDeprecation($node->getName(), $node->getPath());
+                $schema->deprecationMessage = ($deprecation['package'] || $deprecation['version'] ? "Since {$deprecation['package']} {$deprecation['version']}: " : '').$deprecation['message'];
```
This can be simplified after rebase thanks to https://github.com/symfony/symfony/pull/61799
```suggestion
                $schema->deprecationMessage = $node->getDeprecationMessage();
```

GromNaN reviewed (changes requested):
We are making changes in other PRs to unlock this feature.
